lib/google/apis/core/http_command.rb

```diff
@@ -323,10 +333,66 @@ def allow_form_encoding?
 
         private
 
+        def opencensus_begin_span
+          return unless OPENCENSUS_AVAILABLE && options.use_opencensus
+          return if @opencensus_span
+          return unless OpenCensus::Trace.span_context
+
+          @opencensus_span = OpenCensus::Trace.start_span url.path.to_s
+          @opencensus_span.kind = OpenCensus::Trace::SpanBuilder::CLIENT
+          @opencensus_span.put_attribute "http.host", url.host.to_s
+          @opencensus_span.put_attribute "http.method", method.to_s.upcase
+          @opencensus_span.put_attribute "http.path", url.path.to_s
+          if body.respond_to? :bytesize
+            @opencensus_span.put_message_event \
+              OpenCensus::Trace::SpanBuilder::SENT, 1, body.bytesize
+          end
+
+          formatter = OpenCensus::Trace.config.http_formatter
+          if formatter.respond_to? :header_name
+            header[formatter.header_name] = formatter.serialize @opencensus_span.context.trace_context
+          end
+        end
+
+        def opencensus_end_span
+          return unless OPENCENSUS_AVAILABLE
+          return unless @opencensus_span
+          return unless OpenCensus::Trace.span_context
+
+          if @http_res.body.respond_to? :bytesize
+            @opencensus_span.put_message_event \
+              OpenCensus::Trace::SpanBuilder::RECEIVED, 1, @http_res.body.bytesize
+          end
+          status = @http_res.status.to_i
+          if status > 0
+            @opencensus_span.set_status map_http_status status
+            @opencensus_span.put_attribute "http.status_code", status
+          end
+
+          OpenCensus::Trace.end_span @opencensus_span
+          @opencensus_span = nil
+        rescue StandardError => e
+          nil
+        end
+
         def form_encoded?
           @form_encoded
         end
 
+        def map_http_status http_status
+          case http_status
+          when 200..399 then 0 # OK
+          when 401 then 16 # UNAUTHENTICATED
+          when 403 then 7 # PERMISSION_DENIED
+          when 404 then 5 # NOT_FOUND
+          when 429 then 8 # RESOURCE_EXHAUSTED
+          when 400..499 then 3 # INVALID_ARGUMENT
```

Comment:
Just curious, why is the rest of the 400 range given `3` instead of falling through to the default `2`, yet the rest of the 500 range is not similarly mapped? I don't see any advice in [Mapping from HTTP status codes to Trace status codes](https://github.com/census-instrumentation/opencensus-specs/blob/master/trace/HTTP.md#mapping-from-http-status-codes-to-trace-status-codes) for these ranges.

Reply:
That's a good catch. It looks like the status code is supposed to be UNKNOWN if not explicitly mapped, and the rest of the 4xx range is not mapped so I should not be mapping them to INVALID_ARGUMENT. Fixed.